aiida/cmdline/utils/ascii_vis.py

```diff
@@ -279,25 +279,37 @@ def calc_info(node):
     return string
 
 
-def format_call_graph(calc_node, info_fn=calc_info):
+def format_call_graph(calc_node, max_depth: int = None, info_fn=calc_info):
     """
     Print a tree like the POSIX tree command for the calculation call graph
 
     :param calc_node: The calculation node
+    :param max_depth: Maximum depth of the call graph to print
     :param info_fn: An optional function that takes the node and returns a string
         of information to be displayed for each node.
     """
-    call_tree = build_call_graph(calc_node, info_fn=info_fn)
+    call_tree = build_call_graph(calc_node, max_depth=max_depth, info_fn=info_fn)
     return format_tree_descending(call_tree)
 
 
-def build_call_graph(calc_node, info_fn=calc_info):
-    """Build the call graph of a given node."""
+def build_call_graph(calc_node, max_depth: int = None, info_fn=calc_info) -> str:
+    """Build the call graph of a given node.
+
+    :param calc_node: The calculation node
+    :param max_depth: Maximum depth of the call graph to build. Use `None` or -1 for unlimited.
+    :param info_fn: An optional function that takes the node and returns a string
+        of information to be displayed for each node.
+    """
+    if max_depth is None:
+        max_depth = -1
+    if max_depth == 0:
+        return ''
+
     info_string = info_fn(calc_node)
     called = calc_node.called
     called.sort(key=lambda x: x.ctime)
-    if called:
-        return info_string, [build_call_graph(child, info_fn) for child in called]
+    if called and max_depth != 1:
+        return info_string, [build_call_graph(child, max_depth=max_depth - 1, info_fn=info_fn) for child in called]
```

Comment:
If `max_depth` is `-1`, then the function will be then called with `-2`, `-3` ... lead to an infinity iteration?

Reply:
No, if `max_depth` is `-1`, then the stopping condition is given by the finiteness of the workflow as before.

Reply:
Where is this `-1` condition handled? there are no other functions called.

Reply:
The `-1` condition does not need to be handled.
The stopping condition is implicit in the finiteness of the call graph (`called` will eventually be empty).
Just look at the code before this commit.

Reply:
Sorry, I see what you mean ;-)

Reply:
I'll add a test for `-1` as well

Reply:
ah wait, max depth `-1` should actually print everything, so I confused myself ;-)

I will add a test with `0` to prove that the output is empty.

Reply:
I see your point, I think it will be more clear if the `elif` used here to exclusively separate all conditions?
```
if max_depth is None or max_depth == -1:
    return info_string
elif max_depth == 0:
    return ''
else:
  # build_call_graph()
```

Reply:
Hi @unkcpz, your proposed code is not correct.
If `max_depth` is `None` (or `-1`), the whole function body needs to be evaluated.

It would be possible to deal with positive values of `max_depth` + `None` as well, one would then just need to introduce a special case in the recursive call (i.e. reduce `max_depth` is only reduced if it is not `None`).

I personally consider the code clean as it is, but I'll change it and then you can tell me if it is less confusing to you.

Reply:
I've made the change. Not sure whether the code is more elegant this way, but there you go

Reply:
Thanks a lot @ltalirz. It is more clear to me now.

Reply:
Wait, I look at you previous version where there is one layer of `if`.

Reply:
But all good. I totally understand the logic now.

Reply:
It is all very clean after catch `raise ValueError('max_depth must be >= 0')`.